### ISSUE TITLE
Question: masks

### ISSUE THREAD

**Issue by pfeatherstone** [lucidrains/recurrent-memory-transformer-pytorch#15]
So I added some code to plot the attention mask. I use the following:
```
def lengths_to_padding_mask(T: int, lengths: torch.Tensor):
    return torch.arange(T, device=lengths.device).unsqueeze(0) < lengths.unsqueeze(-1)

net = RecurrentMemoryTransformer(
    seq_len=1024,
    num_tokens=259,
    num_memory_tokens=128,
    dim=512,
    depth=1,
    causal=True,
    heads=4,
    dim_head=128,
    use_flash_attn=True,
    rotary_pos_emb=True
).eval()

x = torch.randint(0, 256, (8, 1024))
l = torch.full((x.shape[0],), fill_value=800)
m = lengths_to_padding_mask(x.shape[1], l)
l, mems, _ = net(x, mask=m)
l, mems, _ = net(x, mems, mask=m)
```
and added 
```
import matplotlib.pyplot as plt
img = mask[0,0]
plt.imshow(img)
plt.show()
```
at https://github.com/lucidrains/recurrent-memory-transformer-pytorch/blob/976af1ecba5c1b421a4e204aab31b297c4302b81/recurrent_memory_transformer_pytorch/recurrent_memory_transformer.py#L368

On both forward passes I get the following attention masks:
![mask1_noreadmems](https://github.com/lucidrains/recurrent-memory-transformer-pytorch/assets/45853521/2688866d-e67e-42be-92f7-e126f050be32)

![mask2_readmems](https://github.com/lucidrains/recurrent-memory-transformer-pytorch/assets/45853521/9a6c661a-1015-421e-bf68-950f4c2c4c51)

@lucidrains Is this correct ?

**Comment by lucidrains** [comment]
yea that is correct on first glance

**Comment by pfeatherstone** [comment]
This https://github.com/lucidrains/recurrent-memory-transformer-pytorch/blob/3be7d43604c6921a7dbdc68f88c7f3c534f82d2a/recurrent_memory_transformer_pytorch/recurrent_memory_transformer.py#L371-L372 is really clever!
It's not intuitive to me right now, but I'm sure it should be

**Comment by lucidrains** [comment]
> This
> 
> https://github.com/lucidrains/recurrent-memory-transformer-pytorch/blob/3be7d43604c6921a7dbdc68f88c7f3c534f82d2a/recurrent_memory_transformer_pytorch/recurrent_memory_transformer.py#L371-L372
> 
> is really clever!
> It's not intuitive to me right now, but I'm sure it should be

yea, it was a little jigsaw puzzle
